nav2_controller/src/controller_server.cpp

```diff
@@ -599,10 +599,24 @@ ControllerServer::dynamicParametersCallback(std::vector<rclcpp::Parameter> param
   std::lock_guard<std::mutex> lock(dynamic_params_lock_);
```

Comment:
But you're still in the mutex lock here - this is where the block is from

Reply:
I know, but using try_lock for the parameter changes prevents the deadlock and the parameter changes just fail with an error message, while the controller keeps on running, instead of crashing.

Reply:
I'm not sure how exactly ros2 handles switching between the callbacks, but this is tested in simulation and leads to the behavior I described in my answer above. If I send out parameter changes with the ros2 cli while the controller is currently running I get a failure with the error message I provided and there is no deadlock anymore.

Reply:
It shouldn't though, and I think we're not quite connecting on the issue.  Consider the following 

- In L599 we `lock_guard` the mutex. https://en.cppreference.com/w/cpp/thread/lock_guard This line by itself is waiting for and capturing the mutex in the scope that it exists in.
- This PR wasn't modifying the `lock_guard` so its still being locked immediately once the callback is triggered, for the duration of the callback
-  When you do `dynamic_params_lock_.try_lock()` in L612, you've already locked and have acquired the muted from the `lock_guard` so this will always be successful. In fact the documentation for `try_lock()` states "If try_lock is called by a thread that already owns the mutex, the behavior is undefined. " https://en.cppreference.com/w/cpp/thread/mutex/try_lock so this is probably not correct

I think you need to remove the lock guard and replace it with the `try_lock()` call first thing in the callback. If it fails, exit. If it succeeds, proceed for the callback and then release the lock at the end of the callback.

Reply:
Sry, was on vacation. That the lock_guard in L599 was still in there was indeed a mistake.Taking it out however still leads to the same behavior.
I also changed the lock_guard in the computeControl() function to try_lock/unlock with an error message. Still getting the same behavior in the simulation, so that is the current version.